### PR TITLE
AWS: add parser for aws-cloudtrail events

### PR DESCRIPTION
**Pull request by squioc**
Add the parser and smart-description for the aws-cloudtrail

### CODE REVIEW

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      properties:
+        input_field: '{{json_event.message.eventTime}}'
+        output_field: datetime
+  - name: parsed_source
+    external:
+      name: grok.match
+      properties:
+        input_field: json_event.message.sourceIPAddress
+        output_field: source
+        pattern: '(%{IP:ip}|%{HOSTNAME:domain})'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:19]: Nice trick!

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          source.domain: '{{parsed_source.source.domain}}'
+          action.type: '{{json_event.message.eventType}}'
+          action.name: '{{json_event.message.eventName}}'
+          action.outcome: 'success'
+          action.target: 'network-traffic'
+          user_agent.original: '{{json_event.message.userAgent}}'
+          user.id: '{{json_event.message.userIdentity.accountId}}'
+      - set:
+          user.name: '{{json_event.message.responseElements.user.userId}}'
+        filter: '{{json_event.message.get("responseElements") != None}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:55]: isn't this kind of filter unnecessary now ?

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      - set:
+          tls.cipher: '{{json_event.message.tlsDetails.cipherSuite}}'
+          tls.version: '{{json_event.message.tlsDetails.tlsVersion}}'
+        filter: '{{json_event.message.get("tlsDetails") != None}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:59]: isn't this kind of filter unnecessary now ?

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  set_action_fields:
+    actions:
+      - set:
+          action.properties.errorCode: '{{json_event.message.errorCode}}'
+          action.properties.errorMessage: '{{json_event.message.errorMessage}}'
+          action.properties.recipientAccountId: '{{json_event.message.recipientAccountId}}'
+          action.properties.userIdentity: '{{json_event.message.userIdentity}}'
+          action.properties.resources: '{{json_event.message.resources}}'
+          action.properties.requestParameters.userData: '{{json_event.message.requestParameters.userData}}'
+          action.properties.responseElements.publiclyAccessible: '{{json_event.message.responseElements.publiclyAccessible}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:84]: (suggestion) Maybe, you could have written:
```suggestion
          action.properties: '{{json_event.message}}'
```

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      - set:
+          action.properties.responseElements.pendingModifiedValues.masterUserPassword: '{{json_event.message.responseElements.pendingModifiedValues.masterUserPassword}}'
+        filter: '{{json_event.message.get("responseElement", {}).get("pendingModifiedValues") != None}}'
+
+  set_aws_fields:
+    actions:
+      - set:
+          aws.cloudtrail.event_version: '{{json_event.message.eventVersion}}'
+          aws.cloudtrail.recipient_account_id: '{{json_event.message.recipientAccountId}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:93]: Looks likes some fields are duplicated

For example:
```yaml
aws.cloudtrail.recipient_account_id: '{{json_event.message.recipientAccountId}}'
action.properties.recipientAccountId: '{{json_event.message.recipientAccountId}}'
```

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          event.type: ['access']
+          event.dataset: 'cloudtrail'
+          event.action: '{{json_event.message.eventName or json_event.message.insightDetails.eventName}}'
+          event.code: '{{json_event.message.errorCode or json_event.message.insightDetails.errorCode}}'
+          event.reason: '{{json_event.message.errorMessage}}'
+          event.provider: '{{json_event.message.eventSource or json_event.message.insightDetails.eventSource}}'
+          cloud.provider: 'aws'
+          cloud.service.name: 'cloudtrail'
+          cloud.region: '{{json_event.message.awsRegion}}'
+          cloud.account.id: '{{json_event.message.userIdentity.accountId}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:40]: Duplicated with user.id

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+          json_event.message.responseElements.ConsoleLogin: event.type
+        filter: '{{json_event.message.eventName == "ConsoleLogin"}}'
+
+  set_action_fields:
+    actions:
+      - set:
+          action.properties.errorCode: '{{json_event.message.errorCode}}'
+          action.properties.errorMessage: '{{json_event.message.errorMessage}}'
+          action.properties.recipientAccountId: '{{json_event.message.recipientAccountId}}'
+          action.properties.userIdentity: '{{json_event.message.userIdentity}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:81]: duplicated with `user.id`

File: AWS/aws-cloudtrail/ingest/parser.yml [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      - set:
+          action.properties.responseElements.pendingModifiedValues.masterUserPassword: '{{json_event.message.responseElements.pendingModifiedValues.masterUserPassword}}'
+        filter: '{{json_event.message.get("responseElement", {}).get("pendingModifiedValues") != None}}'
+
+  set_aws_fields:
+    actions:
+      - set:
+          aws.cloudtrail.event_version: '{{json_event.message.eventVersion}}'
+          aws.cloudtrail.recipient_account_id: '{{json_event.message.recipientAccountId}}'
+          aws.cloudtrail.user_identity: '{{json_event.message.userIdentity}}'
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/ingest/parser.yml:94]: duplicate with `user.id`

File: AWS/aws-cloudtrail/tests/event_iam.json [hunk 9]
```diff
@@ -0,0 +1,89 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_iam.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_lambda.json [hunk 10]
```diff
@@ -0,0 +1,126 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_lambda.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_s3.json [hunk 11]
```diff
@@ -0,0 +1,86 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_s3.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_signin.json [hunk 12]
```diff
@@ -0,0 +1,84 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_signin.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_sts.json [hunk 13]
```diff
@@ -0,0 +1,86 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_sts.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/7602ff70-7e5f-42e9-86b2-36803df39183.json [hunk 14]
```diff
@@ -0,0 +1,110 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/7602ff70-7e5f-42e9-86b2-36803df39183.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_cloudtrail.json [hunk 15]
```diff
@@ -0,0 +1,100 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_cloudtrail.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

File: AWS/aws-cloudtrail/tests/event_ec2-error.json [hunk 16]
```diff
@@ -0,0 +1,91 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect_uuid": "9adab001-2187-4f62-8439-367f38b9ba20",
```
Comment (PierrickV) [inline AWS/aws-cloudtrail/tests/event_ec2-error.json:5]:
```suggestion
        "dialect_uuid": "d3a813ac-f9b5-451c-a602-a5994544d9ed",
```

**Review by PierrickV — COMMENTED** [review]
Wrong `dialect_uuid` in test files ?

File: AWS/aws-cloudtrail/ingest/parser.yml (hunk 5 above)
Comment (squioc) [reply AWS/aws-cloudtrail/ingest/parser.yml:93]: `action.properties` is available for retro-compatibility purpose. However, the goal is to remove `action.properties` at the end.

File: AWS/aws-cloudtrail/ingest/parser.yml (hunk 6 above)
Comment (squioc) [reply AWS/aws-cloudtrail/ingest/parser.yml:40]: Applying ECS standard. The account id should landed in `cloud.account.id`

File: AWS/aws-cloudtrail/ingest/parser.yml (hunk 7 above)
Comment (squioc) [reply AWS/aws-cloudtrail/ingest/parser.yml:81]: `action.properties` is available for retro-compatibility purpose

File: AWS/aws-cloudtrail/ingest/parser.yml (hunk 4 above)
Comment (squioc) [reply AWS/aws-cloudtrail/ingest/parser.yml:80]: We have to control the fields produced by the parser. I prefer to specify each field to copy

Comment (gaelmuller) [reply AWS/aws-cloudtrail/ingest/parser.yml:80]: FYI if `action.properties` is declared as a `dict` field, the produced fields are controlled. Only `object` fields are bypassing validation.

**Review by PierrickV — APPROVED** [review]
Looks good to me ! :rocket:

**Pull request merged**